### DISCUSSION TITLE
Unpacking bindvalue inside bind_expression

### DISCUSSION THREAD

**Discussion by amit-hazan** [sqlalchemy/sqlalchemy#9584] [Usage Questions] [answered]
I want to insert a raster data into a raster column in my PostgreSQL DB (+PostGIS). I was trying to build a custom type for that. This type was built for inserting raster data given by memoryview input. Here is my type definition:

```python
import rasterio
from sqlalchemy.types import UserDefinedType

class Raster(UserDefinedType):
    def get_col_spec(self):
        return "RASTER"

    def bind_processor(self, dialect):
        def process(value):
            srid = -1
            with rasterio.io.MemoryFile(value) as memfile:
                with memfile.open() as dataset:
                    if dataset.crs:
                        srid = dataset.crs.to_epsg()
            return value, srid
        return process
    
    def bind_expression(self, bindvalue):
        return func.ST_FromGDALRaster(bindvalue, type_=self)
``` 

According to [RT_ST_FromGDALRaster](https://postgis.net/docs/RT_ST_FromGDALRaster.html), the function takes two arguments. The first argument is the memoryview input and the second argument, srid, can be extracted from the memoryview input data. I did the extraction in bind_processor. The problem is that I didn't find any way to unpack the returned tuple inside bind_expression to get both arguments (bindvalue encapsulates the tuple and I try to do something like *bindvalue), so the above code throws an exception. How can I solve this?

**Comment by zzzeek** [comment]
unless you can come up with a SQL expression that performs the extraction, there's not much other way around this, the constraints are clear: there's one bound value only.   you might want to look at [geoalchemy2](https://geoalchemy-2.readthedocs.io/en/latest/) which has to solve this problem quite often and might have an approach for this specific function already.

**Comment by amit-hazan** [comment, marked as answer]
Thanks for your response!
I finally found a solution to my problem. You can view my discussion here: [https://github.com/geoalchemy/geoalchemy2/issues/437](https://github.com/geoalchemy/geoalchemy2/issues/437)

**Reply by zzzeek** [reply]
great, they work on this problem all the time
